### ISSUE TITLE
[Llama-3.3-70B Galaxy] pcc-80L test reports near-zero / negative PCC but does not assert

### ISSUE THREAD

**Issue by mtairum** [tenstorrent/tt-metal#44367]
## Symptom

The Tier-1 e2e test `models/demos/llama3_70b_galaxy/demo/text_demo.py::test_demo_text[…performance-pcc-80L]` reports essentially-zero or negative PCC against the torch reference and still completes without raising. The latest scheduled run on `main` is the clearest example:

[run 25839608896 — job 75922576110, step 6, line 654](https://github.com/tenstorrent/tt-metal/actions/runs/25839608896/job/75922576110#step:6:654) (`main` @ `c63bb84`):

```
03:35:56  INFO  ... test_demo_text:1073 - PCC: -0.009891958732747128
03:36:26  INFO  ... test_demo_text:1217 - PCC:  0.059200600483592330
03:36:26  INFO  ... test_demo_text:1217 - PCC: -0.026858742676976000
03:36:28  INFO  ... test_demo_text:1217 - PCC:  0.008198889048223404
03:36:29  INFO  ... test_demo_text:1217 - PCC:  0.005213660198256105
03:36:29  INFO  ... test_demo_text:1217 - PCC: -0.013911340886894845
```

A PCC near zero against the reference model means the TT output is uncorrelated with the expected output — i.e. the model is effectively producing noise — yet the test happily continues.

## Things to investigate

1. **Why is the PCC bad?** Identify which stage / layer is corrupting the decode path on Galaxy. (Prefill PCC in the same run is ~0.9+, so the regression appears to be on the decode side.)
2. **Why isn't the test asserting?** The PCC values are logged via `logger.info(...)` but there's no `assert pcc > threshold` gating the test. Add an assertion so a near-zero PCC fails the job instead of silently passing.

## Recommended path forward

This `pcc-80L` test should be **deprecated** in favour of the more robust teacher-forcing test variant (`performance-ci-token-matching`), which compares token-by-token under teacher forcing and surfaces drift earlier and more reliably.

**Action: prioritise adding `performance-ci-token-matching` for Llama-3.3-70B on Galaxy.** Once that is in CI and stable, retire the `pcc-80L` configuration entirely.

## Owner

Open for triage — likely a Llama-Galaxy owner.

**Comment by yieldthought** [comment]
@yieldthought autodebug

**Comment by yieldthought** [comment]
✅ AutoDebug top finding: The pcc-80L decode test corrupts teacher forcing by reusing a stale token, causing near-zero PCCs.

<details>
<summary>Full report</summary>

# AutoDebug Report: tt-metal issue 44367

Issue: https://github.com/tenstorrent/tt-metal/issues/44367

Checkout: `c63bb84`

This was an inspection-only investigation. I did not run the Galaxy repro or any hardware-dependent tests.

## Headline Findings

### 1. The `pcc-80L` decode PCC path corrupts its own teacher-forced sequence

The strongest root cause for the near-zero decode PCCs is in the test harness, not in a specific transformer layer.

`pcc-80L` is the batch-32, 80-layer, `pcc_check=True`, `apc_test=False` parametrization in `models/demos/llama3_70b_galaxy/demo/text_demo.py:584-604`, with id `pcc-80L` at `text_demo.py:666`. In this mode, decode tracing is explicitly disabled:

- `text_demo.py:1150-1153`: `is_enable_trace = enable_trace if not pcc_check else False`
- `generator.py:1122-1149`: the non-traced path reloads host `tokens` and `current_pos` each iteration via `prepare_inputs_decode(...)`

The loop initializes the first decode token from the reference:

- `text_demo.py:1122-1126`: `out_tok = ref_tokens[max_encoded_prompt_len]`, repeated to batch 32

But it only applies later teacher forcing after launching the next decode, and only under `if iteration > 0`:

- `text_demo.py:1159-1174`: launches `generator.decode_forward(out_tok, current_pos, ..., async_read=True, tt_out_logits_saved=...)`
- `text_demo.py:1194-1206`: after the launch, synchronizes/reads the previous output and updates `out_tok`
- `text_demo.py:1198-1199`: teacher forcing sets `out_tok = ref_tokens[max_encoded_prompt_len + iteration + 1]`
- `text_demo.py:1294-1295`: increments `current_pos` and `iteration`

Concrete ledger, with `P = max_encoded_prompt_len`:

| Loop point | `current_pos` | token passed into non-traced decode | Expected reference token for that position |
| --- | ---: | --- | --- |
| before iteration 0 | `P` | `ref_tokens[P]` | `ref_tokens[P]` |
| iteration 0 launch | `P` | `ref_tokens[P]` | correct |
| after iteration 0 | `P + 1` | still `ref_tokens[P]` | not updated because `iteration > 0` was false |
| iteration 1 launch | `P + 1` | `ref_tokens[P]` again | `ref_tokens[P + 1]` |
| after iteration 1 | `P + 2` | set to `ref_tokens[P + 2]` | `ref_tokens[P + 1]` was skipped |

The first corrupted device-side state update is therefore the decode attention KV-cache update, not the final PCC comparison:

- `llama_model.py:709-719`: embeds the supplied token and runs decode forward, then increments device-side decode positions
- `llama_attention.py:509-511`: `paged_fused_update_cache(... update_idxs_tensor=current_pos ...)` writes K/V for the supplied token at the supplied position

At iteration 1, this writes K/V for the stale token `ref_tokens[P]` into cache position `P + 1`. From that point onward the KV cache no longer matches the torch reference sequence. The later iterations may use the right token index again, but the cache is already missing the `ref_tokens[P + 1]` state and contains a duplicate/stale token instead.

This also explains why the first decode token check can look healthy in APC while `pcc-80L` decode PCC is bad:

- `apc_test` stores `tt_out_logits_saved_iter_0` at `text_demo.py:1177-1178`
- at `iteration == 1`, APC compares that iteration-0 tensor against `torch_output[1]` at `text_demo.py:1208-1215`
- iteration 0 used the correct token `ref_tokens[P]`
- `pcc-80L` skips checking iteration 0 and starts comparing at `iteration > 0`, after the stale-token launch has already happened

I also parsed the local reference `.refpt` ZIP storages without `torch` available in this venv. For the 80L reference file, the reference-only PCC between saved step 0 and step 2 is `0.055150865177794724`, close to the first reported bad decode PCC `0.059200600483592330`. That is consistent with a one-step/skip misalignment producing apparently random decode PCCs. This is supporting evidence, not a hardware proof.

### 2. The bad PCCs do not fail CI because `pcc-80L` only logs `comp_pcc`

`comp_pcc` returns `(does_pass, pcc_message)` and does not assert by itself (`models/common/utility_functions.py:521-575`). In `pcc-80L`, the returned boolean is only logged.

Prefill:

- `text_demo.py:1067-1075`: computes and logs PCC
- `text_demo.py:1077-1083`: the assert is gated by `if apc_test`
- `pcc-80L` has `apc_test=False`, so it skips the assert

Decode:

- `text_demo.py:1214-1221`: computes and logs PCC
- `text_demo.py:1223-1229`: the assert is again gated by `if apc_test`
- `pcc-80L` has `apc_test=False`, so it skips the assert

Top-1 and Top-5 are also only logged in this path (`text_demo.py:1231-1240`) and later recorded as measurements; they are not pass/fail gates.

CI currently runs exactly this non-asserting configuration in Tier 1:

- `tests/pipeline_reorg/models_e2e_tests.yaml:125-134`: `pytest --timeout 1000 models/demos/llama3_70b_galaxy/demo/text_demo.py -k "pcc-80L"` with `tier: 1`

### 3. The ticket has an important line-number discrepancy

The pasted issue snippet includes:

```text
test_demo_text:1073 - PCC: -0.009891958732747128
```

In this checkout, `text_demo.py:1073` is the prefill PCC logger, while `text_demo.py:1217` is the decode PCC logger. The ticket text also says prefill PCC in the same run is around 0.9+, so these two observations conflict.

The stale-token decode bug above explains the bad `text_demo.py:1217` decode PCC values. It does not explain a true near-zero prefill PCC at `text_demo.py:1073`. If the line-1073 value is from the exact same checkout/run and is not a log-line mismatch, there is an additional prefill-side problem.

## Other Potential Issues / Lower-Confidence Findings

### Batched prefill saved-logits bookkeeping looks wrong

This is not the headline root cause because `pcc-80L` uses the repeated "This is a test" prompt and the issue narrative says prefill is normally good. It is still a real-looking test/debug bug.

For batch-32 prefill, `generator.py` switches to batched prefill when the batch has uniform 128-token prefill chunks:

- `generator.py:410-417`: `use_batched_prefill = True`
- `generator.py:430-438`: `all_users = [0]`, then `user_id = empty_slots`
- `generator.py:496-504`: passes `user_id=0` and a list of `last_token_idx` values

`process_output_prefill` then splits the output per batch item and loops through all split users:

- `llama_model.py:572-588`: split and per-user LM-head loop
- `llama_model.py:613-617`: after the loop, copies only the final loop's `tt_logits` into `tt_out_logits_saved`

Back in the generator, that one saved tensor is written into `tt_out_logits_all_users[id]` where `id` is just `0` for the batched call:

- `generator.py:543-544`: `tt_out_logits_all_users[id] = tt_out_logits_saved`

So the debug prefill logits checked as user 0 can actually come from the last split processed by the batched loop. With identical prompts this can be masked; with non-identical users this would make the prefill PCC/debug capture misleading.

### The branch diff is only a UMD submodule bump; the TLB memcpy change remains a hardware-sensitive suspect

The top-level diff from `0d83527e14c` to `c63bb84` only changes `tt_metal/third_party/umd`:

```text
Subproject commit 2cf326e3e5216e86dda62df93a729b9298aa8106
Subproject commit 82da9f44fbef1bb1eceae6d4f3be6458a2538103
```

The UMD commits in that range are:

- `bafd6fec Make Device Protocol NOC stateless (#2548)`
- `3b55c7ac Re-land #2357: avoid overlapping memcpy stores on TLB-mapped device memory (#2455)`
- `82da9f44 Add TTDevice::get_io_window (#2541)`

The most suspicious hardware-sensitive delta is the Wormhole TLB-mapped write path changing from a volatile dword loop to SIMD block copies:

- current `tt_metal/third_party/umd/device/pcie/silicon_tlb_window.cpp:118-121`: Wormhole `write_block` calls `memcpy_to_device(...)`
- current `silicon_tlb_window.cpp:204-207`: the destination-aligned middle uses `umd::memcpy_to_device(...)`
- current `device_memcpy.cpp:26-77`: the AVX/SSE stores are explicitly non-volatile intrinsics
- previous `2cf326e3:device/pcie/silicon_tlb_window.cpp:204-210`: the middle copy was a volatile `copy_t` loop

Galaxy is Wormhole, and direct L1/TLB writes are used by dispatch/Fabric setup paths, so this remains a plausible regression-window suspect if a corrected test still shows decode corruption. However, the current `pcc-80L` decode PCC evidence is already contaminated by the test harness token bug. I would not headline the UMD change as the root cause of the reported `text_demo.py:1217` numbers until the teacher-forcing loop is fixed or replaced and the failure still reproduces.

### Other decode-path concerns that did not complete the causal chain

- `DistributedNorm.forward()` ignores the `sharded_output_config` passed into final norm construction (`llama_model.py:98-110`, `distributed_norm.py:60-74`). In practice the LM head reshards its input (`lm_head.py:141-148`), so this looks like a misleading handoff assumption rather than a near-zero PCC cause.
- The decode LM-head all-reduce uses `tt_lm_head_buffer_l1` and deallocates the persistent buffer immediately after enqueueing `all_reduce_async` (`llama_model.py:855-858`, `llama_ccl.py:707-729`). This is riskier than the longer-lived CCL buffers, but I did not find evidence tying it to the reported PCC values.
- I did not find an obvious row/column collective-axis mismatch in the core residual path. The inspected attention, MLP, and LM-head paths use column/row collectives consistently with their sharding contracts.

## Recommended Fixes

1. Fix or remove `pcc-80L` as a correctness signal.

   For the existing code, the minimal decode-loop fix is to set the teacher-forced input for the current position before launching non-traced decode. For `pcc_check and not apc_test`, iteration `i` should feed `ref_tokens[P + i]` at `current_pos = P + i`, then compare the saved logits to `torch_output[i + 1]` when `i + 1 < pcc_decode_len`.

   An even safer short-term option is to disable the async/pipelined token handoff in `pcc_check` mode: launch decode, wait/read/process, update the next teacher-forced token, then launch the next iteration.

2. Make PCC failures assert for `pcc_check`, not only for `apc_test`.

   The non-APC PCC path should assert `does_pass` for both prefill and decode. The APC exact-string target checks can stay separate if the team wants strict drift tracking against `text_demo_targets.json`.

3. Prioritize a Galaxy-compatible `performance-ci-token-matching` run and retire `pcc-80L`.

   The ticket recommendation is sound. Token-matching teacher forcing is a better CI signal than this debug PCC harness because it checks the produced token stream directly and fails on drift. Once it is enabled for Llama-3.3-70B on Galaxy, `pcc-80L` should be removed from Tier 1.

4. After the harness is fixed/replaced, retest the UMD commit range.

   If a corrected token-matching or corrected PCC test still fails only after the UMD bump, focus on commit `3b55c7ac` and the Wormhole TLB-mapped memcpy changes first.

## Verification Notes

- No hardware reproduction was attempted.
- I inspected the Python harness, generator lowering path, decode model path, attention KV update, LM head, CI YAML, and the UMD submodule diff.
- A cheap local static check parsed the stored 80L reference logits and found step-0-vs-step-2 PCC `0.055150865177794724`, consistent with the reported first bad decode PCC magnitude.
- The exact `text_demo.py:1073` prefill log remains unresolved because it conflicts with the ticket narrative that prefill PCC was around 0.9+.

## Continue In Codex

To reopen this investigation thread:

```bash
codex resume 019e504a-3370-7973-be1e-e971e2a59302
```

</details>
